dataflow/src/ops/filter.rs

```diff
@@ -170,47 +170,49 @@ impl Ingredient for Filter {
         &self,
         columns: &[usize],
         key: &KeyType,
+        domains: &DomainNodes,
         states: &'a StateMap,
     ) -> Option<Option<Box<Iterator<Item = Cow<'a, [DataType]>> + 'a>>> {
-        states.get(&*self.src).and_then(|state| {
-            let f = self.filter.clone();
-            let filter = move |r: &[DataType]| {
-                r.iter().enumerate().all(|(i, d)| {
-                    // check if this filter matches
-                    if let Some(ref cond) = f[i] {
-                        match *cond {
-                            FilterCondition::Comparison(ref op, ref f) => {
-                                let v = match *f {
-                                    Value::Constant(ref dt) => dt,
-                                    Value::Column(c) => &r[c],
-                                };
-                                match *op {
-                                    Operator::Equal => d == v,
-                                    Operator::NotEqual => d != v,
-                                    Operator::Greater => d > v,
-                                    Operator::GreaterOrEqual => d >= v,
-                                    Operator::Less => d < v,
-                                    Operator::LessOrEqual => d <= v,
-                                    _ => unimplemented!(),
+        self.lookup(*self.src, columns, key, domains, states)
+            .and_then(|result| {
+                let f = self.filter.clone();
```

Comment:
I don't understand why `f` here has to `clone`. Not introduced by your change though..